File: typescript/src/core/appearance/fonts.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Font registry and text rendering for PDF signature appearances.
 *
 * Available fonts:
 *   - noto-sans: Noto Sans (Google, OFL 1.1) -- default
 *   - ghea-mariam: GHEA Mariam (Armenian)
 *   - ghea-grapalat: GHEA Grapalat (Armenian)
 */

import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export interface FontMetrics {
  name: string;
  unitsPerEm: number;
  ascent: number;
  descent: number;
  capHeight: number;
  bbox: [number, number, number, number];
  stemV: number;
  italicAngle: number;
  defaultWidth: number;
  cmap: Map<number, number>;
  widths: Map<number, number>;
  cidWidthsStr: string;
  tounicodeCmap: string;
  ttfFileName: string;
  /** Resolved absolute path to the TTF file (set during font loading). */
  ttfPath: string;
}

export interface Font {
  name: string;
  textWidth: (text: string, fontSize: number) => number;
  pdfEscape: (text: string) => string;
  metrics: FontMetrics;
}

// -- Font registry ------------------------------------------------------------

export const AVAILABLE_FONTS = ["noto-sans", "ghea-mariam", "ghea-grapalat"] as const;

export type FontName = (typeof AVAILABLE_FONTS)[number];

export const DEFAULT_FONT: FontName = "noto-sans";

interface FontDef {
  metricsModule: string;
  ttfFileName: string;
  baseFontName: string;
}

const FONT_DEFS: Record<FontName, FontDef> = {
  "noto-sans": {
    metricsModule: "./font-data/noto-sans/metrics.js",
    ttfFileName: "NotoSans-Subset.ttf",
    baseFontName: "NotoSans",
  },
  "ghea-mariam": {
    metricsModule: "./font-data/ghea-mariam/metrics.js",
    ttfFileName: "GHEAMariam-Subset.ttf",
    baseFontName: "GHEAMariam",
  },
  "ghea-grapalat": {
    metricsModule: "./font-data/ghea-grapalat/metrics.js",
    ttfFileName: "GHEAGrapalat-Subset.ttf",
    baseFontName: "GHEAGrapalat",
  },
};

const fontCache = new Map<string, Font>();
const defaultFontKey: FontName = DEFAULT_FONT;

const FONT_NAMES: Set<string> = new Set(AVAILABLE_FONTS);

function isValidFontName(name: string): name is FontName {
  return FONT_NAMES.has(name);
}

/**
 * Get a font by key name. Lazy-loads metrics on first access.
 */
export async function getFont(name?: string | null): Promise<Font> {
  const key = name ?? defaultFontKey;
  const cached = fontCache.get(key);
  if (cached) return cached;

  if (!isValidFontName(key)) {
    const available = AVAILABLE_FONTS.join(", ");
    throw new Error(`Unknown font '${key}'. Available: ${available}`);
  }

  return loadFont(key);
}

async function loadFont(key: FontName): Promise<Font> {
  const fontDef = FONT_DEFS[key];

  // Dynamic import of the metrics module
  const mod = await import(fontDef.metricsModule);

  // Resolve absolute path to the TTF file for embedding
  const selfDir = dirname(fileURLToPath(import.meta.url));
  const fontDir = dirname(fontDef.metricsModule);
  const ttfPath = join(selfDir, fontDir, fontDef.ttfFileName);

  const metrics: FontMetrics = {
    name: fontDef.baseFontName,
    unitsPerEm: mod.UNITS_PER_EM,
    ascent: mod.ASCENT,
    descent: mod.DESCENT,
    capHeight: mod.CAP_HEIGHT,
    bbox: mod.BBOX,
    stemV: mod.STEM_V,
    italicAngle: mod.ITALIC_ANGLE,
    defaultWidth: mod.DEFAULT_WIDTH,
    cmap: mod.CMAP,
    widths: mod.WIDTHS,
    cidWidthsStr: mod.CID_WIDTHS_STR,
    tounicodeCmap: mod.TOUNICODE_CMAP,
    ttfFileName: fontDef.ttfFileName,
    ttfPath,
  };

  const font: Font = {
    name: fontDef.baseFontName,
    textWidth: makeTextWidth(metrics),
    pdfEscape: makePdfEscape(metrics),
    metrics,
  };

  fontCache.set(key, font);
  return font;
}

function makeTextWidth(m: FontMetrics): (text: string, fontSize: number) => number {
  const cmap = m.cmap;
  const widths = m.widths;
  const defaultW = m.defaultWidth;
  const upm = m.unitsPerEm;

  return (text: string, fontSize: number): number => {
    let total = 0;
    for (const ch of text) {
      const gid = cmap.get(ch.codePointAt(0) ?? 0) ?? 0;
      total += widths.get(gid) ?? defaultW;
    }
    return (total * fontSize) / upm;
  };
}

function makePdfEscape(m: FontMetrics): (text: string) => string {
  const cmap = m.cmap;
  const questionGid = cmap.get(0x3f) ?? 0; // '?'

  return (text: string): string => {
    const parts: string[] = [];
    for (const ch of text) {
      const gid = cmap.get(ch.codePointAt(0) ?? 0) ?? questionGid;
      parts.push(gid.toString(16).toUpperCase().padStart(4, "0"));
    }
    return `<${parts.join("")}>`;
  };
}

// -- Default font -------------------------------------------------------------

export async function getDefaultFont(): Promise<Font> {
  return getFont(defaultFontKey);
}

// -- Convenience functions ----------------------------------------------------

export async function textWidth(text: string, fontSize: number): Promise<number> {
  const font = await getDefaultFont();
  return font.textWidth(text, fontSize);
}

export async function pdfEscape(text: string): Promise<string> {
  const font = await getDefaultFont();
  return font.pdfEscape(text);
}

export function wrapLines(
  text: string,
  fontSize: number,
  maxWidth: number,
  measure: (text: string, fontSize: number) => number,
): string[] {
  const words = text.split(/\s+/);
  const lines: string[] = [];
  let current = "";

  for (const word of words) {
    if (!word) continue;
    const candidate = current ? `${current} ${word}` : word;
    if (measure(candidate, fontSize) <= maxWidth) {
      current = candidate;
    } else {
      if (current) lines.push(current);
      current = word;
    }
  }
  if (current) lines.push(current);

  return lines;
}

export function encodeTextHex(text: string, cmap: Map<number, number>): string {
  const questionGid = cmap.get(0x3f) ?? 0;
  const parts: string[] = [];
  for (const ch of text) {
    const gid = cmap.get(ch.codePointAt(0) ?? 0) ?? questionGid;
    parts.push(gid.toString(16).toUpperCase().padStart(4, "0"));
  }
  return parts.join("");
}

File: rust/crates/revenant-sign-core/src/appearance/fonts.rs
//! Font registry and text measurement for PDF signature appearances.
//!
//! A [`Font`] bundles the subset-font metrics (glyph coverage, advance widths,
//! the PDF `/W` array and ToUnicode CMap) with the embedded TTF program and the
//! text operations the appearance layer needs: measure width, wrap lines, and
//! encode text as Identity-H hex glyph IDs.
//!
//! Fonts are fully static -- no lazy loading or caching -- so the registry
//! hands out `&'static Font` references, bundling the metrics and the TTF into
//! one type with methods.

use std::fmt;

use crate::appearance::font_data;
use crate::{Result, RevenantError};

/// A subset font: static metrics plus the embedded TTF program.
///
/// Instances are `static`s in the generated `font_data` modules; construct new
/// ones only via the font-preparation pipeline, never by hand.
pub struct Font {
    /// PDF `BaseFont` name, e.g. `"NotoSans"`.
    pub name: &'static str,
    /// Font design units per em (the width denominator).
    pub units_per_em: u32,
    pub ascent: i32,
    pub descent: i32,
    pub cap_height: i32,
    /// FontBBox `(x_min, y_min, x_max, y_max)`.
    pub bbox: (i32, i32, i32, i32),
    pub stem_v: i32,
    pub italic_angle: i32,
    /// Advance width for glyphs absent from [`Font::widths`] (the `/DW` value).
    pub default_width: u32,
    /// Unicode codepoint -> glyph ID, sorted by codepoint.
    pub cmap: &'static [(u32, u16)],
    /// Glyph ID -> advance width (font units), sorted by glyph ID.
    pub widths: &'static [(u16, u16)],
    /// The PDF `/W` array string for the CIDFontType2 dictionary.
    pub cid_widths_str: &'static str,
    /// The ToUnicode CMap stream contents.
    pub tounicode_cmap: &'static str,
    /// The embedded subset TTF (the FontFile2 program).
    pub ttf: &'static [u8],
}

impl fmt::Debug for Font {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Never dump the multi-KB glyph tables or TTF program.
        f.debug_struct("Font")
            .field("name", &self.name)
            .field("units_per_em", &self.units_per_em)
            .field("glyphs", &self.cmap.len())
            .field("ttf_bytes", &self.ttf.len())
            .finish_non_exhaustive()
    }
}

impl Font {
    /// The glyph ID for a Unicode codepoint, if the font's cmap covers it.
    #[must_use]
    pub fn glyph_id(&self, codepoint: u32) -> Option<u16> {
        self.cmap
            .binary_search_by_key(&codepoint, |&(cp, _)| cp)
            .ok()
            .map(|i| self.cmap[i].1)
    }

    /// The advance width (in font units) of a glyph, or [`Font::default_width`].
    #[must_use]
    pub fn advance(&self, glyph_id: u16) -> u32 {
        match self.widths.binary_search_by_key(&glyph_id, |&(g, _)| g) {
            Ok(i) => u32::from(self.widths[i].1),
            Err(_) => self.default_width,
        }
    }

    /// Measure text width in PDF points at a given font size.
    ///
    /// Characters absent from the cmap fall back to glyph 0; the per-glyph
    /// advances are summed as exact integers, then scaled once, so the result
    /// is exact.
    #[must_use]
    pub fn text_width(&self, text: &str, font_size: f64) -> f64 {
        let mut total = 0.0f64;
        for ch in text.chars() {
            let gid = self.glyph_id(u32::from(ch)).unwrap_or(0);
            // Advances are < 2^16 and the sum of a signature line is well below
            // 2^53, so this f64 accumulation is exact.
            total += f64::from(self.advance(gid));
        }
        total * font_size / f64::from(self.units_per_em)
    }

    /// Encode text as concatenated 4-hex-digit glyph IDs for an Identity-H
    /// Type0 font. Characters absent from the cmap use the `?` glyph, so no
    /// invisible `.notdef` glyphs reach the rendered PDF.
    #[must_use]
    pub fn encode_text_hex(&self, text: &str) -> String {
        use std::fmt::Write as _;

        let question = self.glyph_id(u32::from('?')).unwrap_or(0);
        let mut out = String::with_capacity(text.len() * 4);
        for ch in text.chars() {
            let gid = self.glyph_id(u32::from(ch)).unwrap_or(question);
            let _ = write!(out, "{gid:04X}");
        }
        out
    }

    /// Format text as a PDF hex string (`<...>`) ready for the `Tj` operator.
    #[must_use]
    pub fn pdf_escape(&self, text: &str) -> String {
        format!("<{}>", self.encode_text_hex(text))
    }

    /// Word-wrap text to a maximum line width, measuring with this font.
    ///
    /// Splits on whitespace (collapsing runs) and greedily fills lines; a single
    /// word wider than `max_width` still occupies its own line.
    #[must_use]
    pub fn wrap_lines(&self, text: &str, font_size: f64, max_width: f64) -> Vec<String> {
        let mut lines = Vec::new();
        let mut current = String::new();
        for word in text.split_whitespace() {
            let candidate = if current.is_empty() {
                word.to_owned()
            } else {
                format!("{current} {word}")
            };
            if self.text_width(&candidate, font_size) <= max_width {
                current = candidate;
            } else {
                if !current.is_empty() {
                    lines.push(std::mem::take(&mut current));
                }
                word.clone_into(&mut current);
            }
        }
        if !current.is_empty() {
            lines.push(current);
        }
        lines
    }
}

// ── Registry ────────────────────────────────────────────────────────────

/// Font registry keys, in preference order.
pub const AVAILABLE_FONTS: [&str; 3] = ["noto-sans", "ghea-mariam", "ghea-grapalat"];

/// The default font key.
pub const DEFAULT_FONT: &str = "noto-sans";

/// Look up a font by registry key, or the default when `name` is `None`.
///
/// # Errors
///
/// Returns [`RevenantError::Pdf`] if the key is not a known font.
pub fn get_font(name: Option<&str>) -> Result<&'static Font> {
    let key = name.unwrap_or(DEFAULT_FONT);
    match key {
        "noto-sans" => Ok(&font_data::noto_sans::FONT),
        "ghea-mariam" => Ok(&font_data::ghea_mariam::FONT),
        "ghea-grapalat" => Ok(&font_data::ghea_grapalat::FONT),
        _ => Err(RevenantError::Pdf(format!(
            "Unknown font {key:?}. Available: {}",
            AVAILABLE_FONTS.join(", ")
        ))),
    }
}

/// The default font (`noto-sans`).
#[must_use]
pub fn get_default_font() -> &'static Font {
    &font_data::noto_sans::FONT
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_resolves_and_defaults() {
        assert_eq!(get_font(None).unwrap().name, "NotoSans");
        assert_eq!(get_font(Some("noto-sans")).unwrap().name, "NotoSans");
        assert_eq!(get_font(Some("ghea-mariam")).unwrap().name, "GHEAMariam");
        assert_eq!(
            get_font(Some("ghea-grapalat")).unwrap().name,
            "GHEAGrapalat"
        );
        assert_eq!(get_default_font().name, "NotoSans");
    }

    #[test]
    fn rejects_unknown_font() {
        let err = get_font(Some("comic-sans")).unwrap_err();
        assert!(err.to_string().contains("Unknown font"), "{err}");
        assert!(err.to_string().contains("noto-sans"), "{err}");
    }

    #[test]
    fn cmap_and_widths_are_sorted() {
        for key in AVAILABLE_FONTS {
            let font = get_font(Some(key)).unwrap();
            assert!(
                font.cmap.windows(2).all(|w| w[0].0 < w[1].0),
                "{key} cmap not sorted"
            );
            assert!(
                font.widths.windows(2).all(|w| w[0].0 < w[1].0),
                "{key} widths not sorted"
            );
        }
    }

    #[test]
    fn glyph_lookup_matches_generated_cmap() {
        let font = get_default_font();
        // NotoSans cmap maps 'A' (65) -> glyph 34 (0x22).
        assert_eq!(font.glyph_id(u32::from('A')), Some(34));
        assert_eq!(font.encode_text_hex("A"), "0022");
        assert_eq!(font.pdf_escape("A"), "<0022>");
    }

    #[test]
    fn unknown_glyph_encodes_as_question() {
        let font = get_default_font();
        let question_gid = font.glyph_id(u32::from('?')).unwrap();
        // U+1F600 (emoji) is outside the subset -> encodes as the '?' glyph.
        let encoded = font.encode_text_hex("\u{1F600}");
        assert_eq!(encoded, format!("{question_gid:04X}"));
    }

    #[test]
    fn text_width_is_additive_and_scales() {
        let font = get_default_font();
        assert!(font.text_width("", 10.0).abs() < 1e-12);
        let a = font.text_width("A", 10.0);
        let b = font.text_width("B", 10.0);
        let ab = font.text_width("AB", 10.0);
        assert!((ab - (a + b)).abs() < 1e-9);
        // Linear in font size.
        assert!(
            (font.text_width("Signature", 20.0) - 2.0 * font.text_width("Signature", 10.0)).abs()
                < 1e-9
        );
    }

    #[test]
    fn measurements_match_python_reference() {
        // Reference values from the Python client (source of truth), proving the
        // embedded metrics and measurement arithmetic are byte-faithful.
        let noto = get_font(Some("noto-sans")).unwrap();
        assert!((noto.text_width("Signature", 12.0) - 54.684).abs() < 1e-9);
        assert!((noto.text_width("Test 2026 UTC+4", 10.0) - 78.31).abs() < 1e-9);
        assert_eq!(noto.pdf_escape("Sig"), "<0034004A0048>");

        let ghea = get_font(Some("ghea-mariam")).unwrap();
        assert!((ghea.text_width("Ստորագրություն", 10.0) - 78.99).abs() < 1e-9);
        assert_eq!(ghea.pdf_escape("Ա"), "<00C0>");
    }

    #[test]
    fn wrap_lines_greedy() {
        let font = get_default_font();
        // A generous width keeps everything on one line.
        assert_eq!(
            font.wrap_lines("one two three", 10.0, 1000.0),
            vec!["one two three"]
        );
        // A tiny width forces one word per line.
        let narrow = font.wrap_lines("one two three", 10.0, 1.0);
        assert_eq!(narrow, vec!["one", "two", "three"]);
        // Whitespace runs collapse.
        assert_eq!(font.wrap_lines("  a   b  ", 10.0, 1000.0), vec!["a b"]);
    }
}
